File: backend/ivfitter/core/synthetic_trace.py
"""Forward-simulated IV trace generation.

This module wraps the existing model evaluator with sweep construction,
optional noise, optional compliance clipping, and trace metadata. It does not
define a separate equation set for synthetic data.
"""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Literal

import numpy as np
from pydantic import BaseModel, Field

from ivfitter import __version__
from .evaluation import predict_current
from .model_spec import ComponentSpec, ModelSpec

MAX_SYNTHETIC_POINTS = 10000
GENERATOR_VERSION = "synthetic-trace-v1"


class SyntheticNoiseConfig(BaseModel):
    mode: Literal["none", "gaussian_absolute", "gaussian_relative"] = "none"
    noise_level_A: float = 0.0
    relative_noise_fraction: float = 0.0


class SyntheticArtifactConfig(BaseModel):
    compliance_enabled: bool = False
    compliance_current_A: float | None = None


class SyntheticTraceResult(BaseModel):
    trace_name: str
    voltage_V: list[float]
    current_A: list[float]
    metadata: dict


class SyntheticTraceRequest(BaseModel):
    model: ModelSpec
    voltage_start: float
    voltage_stop: float
    voltage_step: float
    noise_config: SyntheticNoiseConfig = Field(default_factory=SyntheticNoiseConfig)
    artifact_config: SyntheticArtifactConfig = Field(default_factory=SyntheticArtifactConfig)
    trace_name: str = "synthetic_trace"
    seed: int | None = None


def _component_label(comp: ComponentSpec) -> str:
    form = comp.evaluation_form or "default"
    placement = comp.placement or "default"
    return f"{comp.id} ({comp.location}/{comp.function_type}, form={form}, placement={placement})"


def _validate_component_support(model: ModelSpec) -> None:
    supported: dict[str, set[str]] = {
        "core": {"diode"},
        "series": {
            "constant_rs",
            "shunt",
            "softplus_rs_modifier",
            "custom",
            "series_diode_barrier",
            "series_power_law_drop",
        },
        "parallel": {
            "shunt",
            "constant_rs",
            "power_law",
            "soft_breakdown",
            "photocurrent_constant",
            "photocurrent_voltage_dependent",
            "custom",
        },
    }
    for location in ("core", "series", "parallel"):
        for comp in getattr(model, location):
            if comp.function_type not in supported[location]:
                raise ValueError(f"Synthetic generation does not support {_component_label(comp)}.")
            if comp.evaluation_form == "implicit_relation":
                raise ValueError(f"Synthetic generation does not support implicit evaluation form for {_component_label(comp)}.")
            if location == "core" and comp.evaluation_form not in (None, "current_branch"):
                raise ValueError(f"Synthetic generation does not support {_component_label(comp)}.")
            if location == "series" and comp.evaluation_form not in (None, "voltage_drop", "conductance_modifier"):
                raise ValueError(f"Synthetic generation does not support {_component_label(comp)}.")
            if location == "parallel" and comp.evaluation_form not in (None, "current_branch"):
                raise ValueError(f"Synthetic generation does not support {_component_label(comp)}.")


def build_voltage_sweep(voltage_start: float, voltage_stop: float, voltage_step: float) -> np.ndarray:
    start = float(voltage_start)
    stop = float(voltage_stop)
    step = float(voltage_step)
    if not np.isfinite(start) or not np.isfinite(stop) or not np.isfinite(step):
        raise ValueError("Voltage start, stop, and step must be finite.")
    if step <= 0:
        raise ValueError("Voltage step must be > 0.")
    if start == stop:
        raise ValueError("Voltage start and stop must be different.")
    span = abs(stop - start)
    count = int(np.floor(span / step + 1e-12)) + 1
    if count < 2:
        raise ValueError("Voltage sweep must contain at least two points.")
    if count > MAX_SYNTHETIC_POINTS:
        raise ValueError(f"Voltage sweep has too many points ({count}); limit is {MAX_SYNTHETIC_POINTS}.")
    direction = 1.0 if stop > start else -1.0
    return start + direction * step * np.arange(count, dtype=float)


def _ground_truth_parameters(model: ModelSpec) -> dict[str, float]:
    out: dict[str, float] = {}
    for group_name in ("core", "series", "parallel"):
        for comp in getattr(model, group_name):
            for name, spec in comp.params.items():
                out[f"{comp.id}.{name}"] = float(spec.value)
    return out


def _add_noise(current: np.ndarray, noise_config: SyntheticNoiseConfig, seed: int | None) -> np.ndarray:
    mode = noise_config.mode
    if mode == "none":
        return current
    rng = np.random.default_rng(seed)
    if mode == "gaussian_absolute":
        sigma = float(noise_config.noise_level_A)
        if sigma < 0 or not np.isfinite(sigma):
            raise ValueError("Absolute noise level must be finite and >= 0.")
        return current + rng.normal(0.0, sigma, size=current.shape)
    if mode == "gaussian_relative":
        fraction = float(noise_config.relative_noise_fraction)
        if fraction < 0 or not np.isfinite(fraction):
            raise ValueError("Relative noise fraction must be finite and >= 0.")
        return current + rng.normal(0.0, np.abs(current) * fraction, size=current.shape)
    raise ValueError(f"Unsupported noise mode {mode!r}.")


def _apply_compliance(current: np.ndarray, artifact_config: SyntheticArtifactConfig) -> np.ndarray:
    if not artifact_config.compliance_enabled:
        return current
    limit = artifact_config.compliance_current_A
    if limit is None or not np.isfinite(limit) or limit <= 0:
        raise ValueError("Compliance current must be finite and > 0 when compliance is enabled.")
    return np.clip(current, -float(limit), float(limit))


def generate_synthetic_trace(
    model: ModelSpec,
    voltage_start: float,
    voltage_stop: float,
    voltage_step: float,
    noise_config: SyntheticNoiseConfig,
    artifact_config: SyntheticArtifactConfig,
    trace_name: str,
    seed: int | None = None,
) -> SyntheticTraceResult:
    _validate_component_support(model)
    voltage = build_voltage_sweep(voltage_start, voltage_stop, voltage_step)
    current = np.asarray(predict_current(voltage, model), dtype=float)
    if current.shape != voltage.shape:
        raise ValueError("Synthetic model evaluation returned a current array with the wrong shape.")
    if not np.all(np.isfinite(current)):
        raise ValueError("Synthetic model evaluation produced non-finite currents; no approximate fallback was used.")
    current = _add_noise(current, noise_config, seed)
    current = _apply_compliance(current, artifact_config)
    if not np.all(np.isfinite(current)):
        raise ValueError("Synthetic trace generation produced non-finite currents.")

    clean_name = trace_name.strip() or "synthetic_trace"
    metadata = {
        "synthetic": True,
        "generator_version": GENERATOR_VERSION,
        "software_version": __version__,
        "model_snapshot": model.model_dump(mode="json"),
        "ground_truth_parameters": _ground_truth_parameters(model),
        "voltage_start": float(voltage_start),
        "voltage_stop": float(voltage_stop),
        "voltage_step": float(voltage_step),
        "noise_config": noise_config.model_dump(mode="json"),
        "artifact_config": artifact_config.model_dump(mode="json"),
        "seed": seed,
        "created_at": datetime.now(timezone.utc).isoformat(),
    }
    return SyntheticTraceResult(
        trace_name=clean_name,
        voltage_V=voltage.tolist(),
        current_A=current.tolist(),
        metadata=metadata,
    )

File: frontend/src/model/syntheticTrace.ts
import type { ModelSpec, SyntheticArtifactConfig, SyntheticNoiseConfig, SyntheticTraceRequest, SyntheticTraceResponse, TraceData } from "./types";

export const MAX_SYNTHETIC_POINTS = 10000;

export interface SyntheticTraceFormState {
  traceName: string;
  voltageStart: string;
  voltageStop: string;
  voltageStep: string;
  noiseMode: SyntheticNoiseConfig["mode"];
  noiseLevelA: string;
  relativeNoiseFraction: string;
  seed: string;
  complianceEnabled: boolean;
  complianceCurrentA: string;
}

export interface SyntheticTraceValidation {
  ok: boolean;
  error: string | null;
  pointCount: number;
}

export const defaultSyntheticTraceForm: SyntheticTraceFormState = {
  traceName: "synthetic_trace",
  voltageStart: "-1",
  voltageStop: "1",
  voltageStep: "0.02",
  noiseMode: "none",
  noiseLevelA: "1e-12",
  relativeNoiseFraction: "0.01",
  seed: "1",
  complianceEnabled: false,
  complianceCurrentA: "0.001",
};

function asNumber(value: string): number {
  return Number(value.trim());
}

export function buildSyntheticTracePayload(form: SyntheticTraceFormState, model: ModelSpec, traceName: string): SyntheticTraceRequest {
  const noise_config: SyntheticNoiseConfig = form.noiseMode === "gaussian_absolute"
    ? { mode: "gaussian_absolute", noise_level_A: Number(form.noiseLevelA) }
    : form.noiseMode === "gaussian_relative"
      ? { mode: "gaussian_relative", relative_noise_fraction: Number(form.relativeNoiseFraction) }
      : { mode: "none" };
  return {
    model,
    trace_name: traceName,
    voltage_start: Number(form.voltageStart),
    voltage_stop: Number(form.voltageStop),
    voltage_step: Number(form.voltageStep),
    noise_config,
    artifact_config: form.complianceEnabled
      ? { compliance_enabled: true, compliance_current_A: Number(form.complianceCurrentA) }
      : { compliance_enabled: false },
    seed: form.seed.trim() ? Number(form.seed) : null,
  };
}

export function syntheticTraceCsv(response: SyntheticTraceResponse, header = "voltage_V,current_A"): string {
  return [header, ...response.voltage_V.map((v, idx) => `${v},${response.current_A[idx]}`)].join("\n");
}

export function syntheticPointCount(voltageStart: number, voltageStop: number, voltageStep: number): number {
  if (!Number.isFinite(voltageStart) || !Number.isFinite(voltageStop) || !Number.isFinite(voltageStep) || voltageStep <= 0 || voltageStart === voltageStop) return 0;
  return Math.floor(Math.abs(voltageStop - voltageStart) / voltageStep + 1e-12) + 1;
}

export function validateSyntheticTraceForm(form: SyntheticTraceFormState): SyntheticTraceValidation {
  const start = asNumber(form.voltageStart);
  const stop = asNumber(form.voltageStop);
  const step = asNumber(form.voltageStep);
  if (!Number.isFinite(start) || !Number.isFinite(stop) || !Number.isFinite(step)) return { ok: false, error: "Voltage start, stop, and step must be finite numbers.", pointCount: 0 };
  if (step <= 0) return { ok: false, error: "V step must be > 0.", pointCount: 0 };
  if (start === stop) return { ok: false, error: "V start and V stop must be different.", pointCount: 0 };
  const pointCount = syntheticPointCount(start, stop, step);
  if (pointCount > MAX_SYNTHETIC_POINTS) return { ok: false, error: `Voltage sweep has ${pointCount} points; limit is ${MAX_SYNTHETIC_POINTS}.`, pointCount };
  if (form.noiseMode === "gaussian_absolute") {
    const level = asNumber(form.noiseLevelA);
    if (!Number.isFinite(level) || level < 0) return { ok: false, error: "Absolute noise level must be >= 0 A.", pointCount };
  }
  if (form.noiseMode === "gaussian_relative") {
    const fraction = asNumber(form.relativeNoiseFraction);
    if (!Number.isFinite(fraction) || fraction < 0) return { ok: false, error: "Relative noise fraction must be >= 0.", pointCount };
  }
  if (form.seed.trim() && !Number.isInteger(Number(form.seed))) return { ok: false, error: "Random seed must be an integer.", pointCount };
  if (form.complianceEnabled) {
    const limit = asNumber(form.complianceCurrentA);
    if (!Number.isFinite(limit) || limit <= 0) return { ok: false, error: "Compliance current must be > 0 A.", pointCount };
  }
  return { ok: true, error: null, pointCount };
}

export function syntheticResponseToTrace(response: SyntheticTraceResponse): TraceData {
  return {
    voltage_V: response.voltage_V,
    current_A: response.current_A,
    trace_id: response.trace_name,
    metadata: {
      ...response.metadata,
      voltage_unit: "V",
      current_unit: "A",
      voltage_unit_factor_to_V: 1,
      current_unit_factor_to_A: 1,
      unit_mode: "import_unit_to_si_internal",
    },
  };
}

export function uniqueTraceId(baseName: string, traces: TraceData[]): string {
  const clean = baseName.trim().replace(/\s+/g, " ") || "synthetic_trace";
  const existing = new Set(traces.map((trace) => trace.trace_id));
  if (!existing.has(clean)) return clean;
  let suffix = 2;
  let next = `${clean} ${suffix}`;
  while (existing.has(next)) next = `${clean} ${++suffix}`;
  return next;
}

export function appendSyntheticTrace(traces: TraceData[], response: SyntheticTraceResponse): { traces: TraceData[]; selectedTraceId: string } {
  const trace = syntheticResponseToTrace(response);
  const traceId = uniqueTraceId(trace.trace_id, traces);
  const nextTrace = traceId === trace.trace_id ? trace : { ...trace, trace_id: traceId, metadata: { ...trace.metadata, dataset_name: traceId } };
  return { traces: [...traces, nextTrace], selectedTraceId: nextTrace.trace_id };
}
